src/dtx/dtx_common.c

```diff
@@ -543,6 +480,214 @@ dtx_conflict(daos_handle_t coh, struct dtx_handle *dth, uuid_t po_uuid,
+	 * current DTX before remote replica(s) modification by race. So
+	 * let's check DTX status locally before marking as 'committable'.
+	 */
+	if (dlh->dlh_handled_time <= cont->sc_dtx_resync_time) {
+		rc = vos_dtx_check_committable(cont->sc_hdl, NULL,
+					       &dth->dth_xid, 0, false);
+		if (rc < 0) {
+			result = (rc == -DER_NONEXIST ? -DER_INPROGRESS : rc);
+			D_GOTO(fail, result);
+		}
+	}
+
+	rc = vos_dtx_add_cos(dth->dth_coh, &dth->dth_oid, &dth->dth_xid,
+			     dth->dth_dkey_hash, dlh->dlh_handled_time,
+			     dth->dth_intent == DAOS_INTENT_PUNCH ?
+			     true : false);
+	if (rc != 0) {
+		D_ERROR(DF_UUID": Fail to add DTX "DF_DTI" to CoS cache: %d. "
+			"Try to commit it sychronously.\n",
+			DP_UUID(cont->sc_uuid), DP_DTI(&dth->dth_xid), rc);
+		dth->dth_sync = 1;
+	}
+
+	if (dth->dth_sync) {
+		rc = dtx_commit(cont_hdl->sch_pool->spc_uuid,
+				cont->sc_uuid, &dth->dth_dte, 1,
+				cont_hdl->sch_pool->spc_map_version);
+		if (rc != 0) {
+			D_ERROR(DF_UUID": Fail to sync commit DTX "DF_DTI
+				": rc = %d\n", DP_UUID(cont->sc_uuid),
+				DP_DTI(&dth->dth_xid), rc);
+			D_GOTO(fail, result = rc);
+		}
+	}
+
+fail:
+	if (result < 0)
+		dtx_abort(cont_hdl->sch_pool->spc_uuid, cont->sc_uuid,
+			  &dth->dth_dte, 1,
+			  cont_hdl->sch_pool->spc_map_version);
+out_free:
+	D_DEBUG(DB_TRACE,
+		"Stop the DTX "DF_DTI" ver %u, dkey %llu, intent %s, "
+		"%s, %s: rc = %d\n",
+		DP_DTI(&dth->dth_xid), dth->dth_ver,
+		(unsigned long long)dth->dth_dkey_hash,
+		dth->dth_intent == DAOS_INTENT_PUNCH ? "Punch" : "Update",
+		dth->dth_sync ? "sync" : "async",
+		dth->dth_non_rep ? "non-replicated" : "replicated", result);
+
+	if (dth->dth_dti_cos != NULL)
+		D_FREE(dth->dth_dti_cos);
+out:
+	return result > 0 ? 0 : result;
+}
+
+/**
+ * Prepare the DTX handle in DRAM.
+ *
+ * XXX: Currently, we only support to prepare the DTX against single DAOS
+ *	object and single dkey.
+ *
+ * \param dti		[IN]	The DTX identifier.
+ * \param oid		[IN]	The target object (shard) ID.
+ * \param coh		[IN]	Container open handle.
+ * \param epoch		[IN]	Epoch for the DTX.
+ * \param dkey_hash	[IN]	Hash of the dkey to be modified if applicable.
+ * \param conflict	[IN]	Hash of the dkey to be modified if applicable.
+ * \param dti_cos	[IN,OUT]The DTX array to be committed because of shared.
+ * \param dti_cos_count [IN,OUT]The @dti_cos array size.
+ * \param pm_ver	[IN]	Pool map version for the DTX.
+ * \param intent	[IN]	The intent of related modification.
+ * \param leader	[IN]	The target (to be modified) is leader or not.
+ * \param dth		[OUT]	Pointer to the DTX handle.
+ *
+ * \return			Zero on success, negative value if error.
+ */
+int
+dtx_begin(struct dtx_id *dti, daos_unit_oid_t *oid, daos_handle_t coh,
+	  daos_epoch_t epoch, uint64_t dkey_hash,
+	  struct dtx_conflict_entry *conflict, struct dtx_id *dti_cos,
+	  int dti_cos_cnt, uint32_t pm_ver, uint32_t intent,
+	  struct dtx_handle *dth)
```

Comment:
Honestly, I more like the current implementation, means the dtx_begin() with @leader parameter, it seems NOT necessary to make two different sets of APIs: dtx_leader_begin() and dtx_begin(). I did not see the benefit for that. If you really want to split leader and non-leader logic, then we can do that inside DTX but keep the dtx_begin() API uniquely.

Reply:
If you compare dtx_leader_begin/end and dtx_begin/end, they are doing completely different thing, so it probably better to separate them.  I thought to combine them, but seems this is better, IMHO.  Since we already decide to separate dtx into distribute and undo log, so I split the function.  I want to do more on dtx, but liang told me you or jeff is working on undo log part. so I will wait.  Anyway, with this patch, I can add EC with this distribute api now.

Reply:
Different parameters, different implementation,  why do you want to do that?  I think 2 api make everything clear.

Reply:
But that double the DTX APIs count. For example, just like the "ls" command with kinds of options, we can re-implement multiple "ls" commands for different options: ls_l, ls_a, ls_i, and so on, each of them is more simple and clear than current compound "ls" command, but from the users' perspective, which one is more preferable? As my understand, merging similar semantic functions into the same API seems more easy to be used.

Reply:
Sorry, for me, two functions for different things seems making more sense.

Reply:
I trend to maintain less APIs that is more easy to be remembered and less confused for users. But since you think it is helpful for the subsequent EC work, then let's hold the argument temporarily and discuss it in further some time later.